### ISSUE TITLE
Admin Styles and Scripts not loading due to wrong path

### ISSUE THREAD

**Issue by bitstarr** [WebDevStudios/CMB2#432]
As said in title, CSS and JS get the wrong path (absolute file system path) on registration.
```
<link rel='stylesheet' id='cmb2-styles-css'  href='http://developer.dev/home/dev/web/wordpress/wp-content/plugins/cmb2/css/cmb2.min.css?ver=4.2.4' type='text/css' media='all' />
```

The problem is, that the ```url``` function in ```CMB2_Utils``` class utilizes ```cmb2_dir()``` and so the ```CMB2_DIR``` constant which is constructed for for includes.

I replaced the line
```
$this->url = trailingslashit( apply_filters( 'cmb2_meta_box_url', set_url_scheme( $cmb2_url ), CMB2_VERSION ) );
```
with
```
$this->url = plugins_url( '', __FILE__ ) . '/../';
```
as a workaround.

Please find a better solution.

**Comment by jtsternberg** [comment]
Please don't hack the plugin as you could break it for other plugins/themes depending on it. There is documentation for updating that url if it is not working properly: https://github.com/WebDevStudios/CMB2/wiki/Troubleshooting

TL;DR: there is a filter in place so hacking the plugin is not necessary.

**Comment by gyrus** [comment]
I'm having the issue with CMB2 symlinked locally in my dev environment. My workaround:

    add_filter( 'cmb2_meta_box_url', 'pilau_cmb2_meta_box_url' );
    function pilau_cmb2_meta_box_url( $url ) {
        if ( defined( 'WP_LOCAL_DEV' ) && WP_LOCAL_DEV ) {
            $url = plugins_url( 'cmb2/' );
        }
        return $url;
    }

I wonder why `plugins_url()` isn't used in the first place?

**Comment by jtsternberg** [comment]
Because CMB2 can be loaded as a plugin, from a plugin, from mu-plugins, from a theme, etc. 

I would like to solve this a bit better though. Can you provide more insight into your file structure and where CMB2 is loaded?

**Comment by gyrus** [comment]
Ah, yes, forgot it could be packaged with a theme.

I have all my own plugins, and any I may contribute to, in a local folder outside the web root, /www-projects/, cloned from GitHub. I then just symlink those plugins locally. I can see there's a problem if you have to cater for the different locations CMB2 may be in - and maybe the filter is the best or only option? That's working for me for now, so no urgency, but interested if you find a solution that covers all bases.

**Comment by gyrus** [comment]
I didn't realise this had happened, I guess I only started symlinking plugins after 3.9. But maybe of use:

https://make.wordpress.org/core/2014/04/14/symlinked-plugins-in-wordpress-3-9/

**Comment by MickeyKay** [comment]
Similar issue here. We're including a bunch of composer dependencies in our WP root, and when I require `init.php` from our theme's functions.php file with something like this:
```require( ABSPATH . '/vendor/webdevstudios/cmb2/init.php' );```

. . . I end up with a console error with CMB2 attempting to pull in the following file (notice the bogus path):
`{our_root_url}/srv/{our_root_path}/vendor/webdevstudios/cmb2/js/cmb2.min.js?ver=2.2.0`

I know CMB2 is meant to be used as a standalone library, however it seems like it is required to live within wp-content for any of this to work, which ends up not working so well for the use case I'm describing.

Any ideas on this one?

**Comment by jtsternberg** [comment]
@MickeyKay Yep, just need to filter that URL. https://github.com/WebDevStudios/CMB2/wiki/Troubleshooting#cmb2-urls-issues (like [mentioned above](https://github.com/WebDevStudios/CMB2/issues/432#issuecomment-132271801)).

**Comment by jtsternberg** [comment]
This issue should be solved for almost all use-cases now (without need to filter). If any of you are able to test with the `trunk` branch, please do so.

**Comment by justnorris** [comment]
Tried trunk, doesn't help on my local development environment.

**Comment by jtsternberg** [comment]
@justnorris yes, the filter is there for the scenarios where CMB2 is not smart enough to figure it out. I've updated the documentation though, as there are some caveats to consider when using the filter (namely, try not to break CMB2 for other implementors): https://github.com/WebDevStudios/CMB2/wiki/Troubleshooting#cmb2-urls-issues

Can you provide me the broken URL vs the correct URL for your install? I want to see if there's ways to smarten CMB2's logic a bit more.

**Comment by justnorris** [comment]
@jtsternberg Github didn't notify me, sorry for the late response.

I'm running a local MAMP, with an actual domain, pointing to 10.0.0.10 ( my local IP assigned to my computer by the router ).

So I get 404 that looks like this:
```
http://bird.localbox.in/Users/N/Projects/Colormelon/Plugins/portfolio-plugin/vendor/webdevstudios/cmb2/css/cmb2.css?ver=4.6.1 Failed to load resource: the server responded with a status of 404 (Not Found)
```

**Comment by tw2113** [comment]
Honestly, URLs for these types of resources should probably utilize the appropriate functions to get browser friendly URLs, and not server path versions.

In the meatime, @justnorris could use https://github.com/WebDevStudios/CMB2/wiki/Troubleshooting#cmb2-urls-issues

and have it return the following, if I'm seeing things correctly.

```
return str_replace( '/Users/N/Projects/Colormelon/Plugins/', '/wp-content/plugins/', $url );
```

**Comment by jtsternberg** [comment]
> Honestly, URLs for these types of resources should probably utilize the appropriate functions to get browser friendly URLs, and not server path versions.

This is not possible since CMB2 can be bundled/included in many places. Believe me, it's been discussed/hashed out before, including the rest of this thread.

**Comment by tw2113** [comment]
fair enough.

**Comment by justnorris** [comment]
This works perfectly for me:
```
if ( defined( "WP_DEBUG" ) && WP_DEBUG ) {
	function update_cmb2_meta_box_url( $url ) {

		return plugin_dir_url( __FILE__ ) . '/vendor/webdevstudios/cmb2';
	}

	add_filter( 'cmb2_meta_box_url', 'update_cmb2_meta_box_url' );
}
```

**Comment by JiveDig** [comment]
Just hit this issue with a (locally symlinked) plugin that bundles CMB2. Just thinking about it... if I run the filter in my plugin, but another plugin bundling CMB2 gets loaded first, won't my filter run and screw things up?

**Comment by tw2113** [comment]
Logically, it would, but at the same time, the file you're referring to with the filter would still exist and load. At that point, the primary potential errors would be js selectors not finding things like expected, or some styles mismatching.

**Comment by JiveDig** [comment]
Interesting note, it only fails on my term metabox. Not on my post/page metabox. The filter fixed it, though i'm leery to use it.

**Comment by tw2113** [comment]
Perhaps conditionally add the filter? Limit the scope of where it gets added.

**Comment by jtsternberg** [comment]
The trick ([as identified in the docs for that filter](https://github.com/CMB2/CMB2/wiki/Troubleshooting#cmb2-urls-issues)) is to do a `string_replace` on the url to only replace the bad bits with good bits, vs explicitly setting the URL to our local version. In my example on the wiki page, it's removing the incorrect bits of the symlink path, but this will not break things if CMB2 is loaded from another plugin/theme/etc. Hopefully that makes sense.

**Comment by taimurahmad971** [comment]
problem solved by just commenting in init.php
                      `  // Kick the whole thing off.
			//require_once( cmb2_dir( 'bootstrap.php' ) );
			//cmb2_bootstrap();`

**Comment by tw2113** [comment]
Given that those two lines are more or less essential to CMB2 running as a whole, it's definitely not a solution by any means.
